SharedPackages/BrowserServicesKit/Sources/Bookmarks/FaviconsFetcher/FaviconsFetchOperation.swift

```diff
@@ -100,21 +100,27 @@ final class FaviconsFetchOperation: Operation, @unchecked Sendable {
     }
 
     func fetchFavicons() async throws {
-        var idsToProcess = try stateStore.getBookmarkIDs()
+        var (bookmarkDomains, idsToProcess) = try await MainActor.run { () -> (BookmarkDomains?, Set<String>) in
+            let idsToProcess = try stateStore.getBookmarkIDs()
 
-        guard !idsToProcess.isEmpty else {
-            Logger.bookmarks.debug("No new Favicons to fetch")
-            return
+            guard !idsToProcess.isEmpty else {
+                Logger.bookmarks.debug("No new Favicons to fetch")
+                return (nil, [])
+            }
+            Logger.bookmarks.debug("Favicons Fetch Operation started")
+
+            var bookmarkDomains = mapBookmarkDomainsToUUIDs(for: idsToProcess)
+            bookmarkDomains.filterDomains { [weak self] domain in
+                self?.faviconStore.hasFavicon(for: domain) == false
```

Comment:
it seems we only need to call `filterDomains`+`hasFavicon` on the main thread, the rest might run in the background?

Reply:
You’re correct! But filterDomains is mutable and I also had errors about passing sending bookmarkDomains between asynchronous contexts. Happy to take suggestions on how to rewrite it so that it makes sense and only filters on the main actor. I suppose we’d have to rewrite BookmarksDomains somehow which seemed a larger change so I decided to defer it for now.